### ISSUE TITLE
IEC to Shugart interface ..

### ISSUE THREAD

**Issue by gilmod714** [dhansel/IECDevice#1]
would it be possible to use this library to interface IEC to a PC floppy 3.5 or 5.25 ?! .. with or without 1541 controller ..

**Comment by dhansel** [comment]
If I understand you correctly then this is exactly what one of my examples for this library does:
https://github.com/dhansel/IECDevice/tree/main/examples/IECFDCMega

**Comment by gilmod714** [comment]
Wow ,  this is awesome!!! .. my scenario is to connect a PC style floppy to IEC bus ... I have c64 & VICE emulator on my PC that connects to my "physical" 1541-drive via using Zoomfloppy adapter (usb->IEC) .. but those 1541's are plagued by having bad heads .. (which was my case recently) .. so I wanted to switch to shugart style drives instead , and ditch the 1541 hardware altogether..

I will give this a try ..

Thank you David, excellent work :)

**Comment by gilmod714** [comment]
Is there a schematic for the Mega-shield somewhere?! .. I can't find it .. only for the pico ..

**Comment by dhansel** [comment]
There really isn't much to those boards, mostly just wiring. Anyways, I have now added links to the schematics here:
https://github.com/dhansel/ArduinoFDC?tab=readme-ov-file#arduinofdc-shields
Direct link for the Mega shield: https://raw.githubusercontent.com/dhansel/ArduinoFDC/main/ArduinoFDCMegaShieldSchematic.pdf

**Comment by gilmod714** [comment]
Thank you David , I appreciate the links .. looks like the 74HC125 buffer isn't needed for the Mega ?! just pullups on read & index?! ..

this one looks different with extra circuitry though.. I think this is what I need to build for the IEC bus

![IECFDCMega](https://github.com/user-attachments/assets/8003607f-21ea-4eda-9eaf-09a33c6dc317)

**Comment by dhansel** [comment]
The 74HC125 is not part of the shield itself (the shield is generic for the ArduinoFDC project) but it is part of this example,
as described in the example's "Wiring" section.

**Comment by gilmod714** [comment]
Ok, I was a bit confused, I also see 2 extra TRN, 3 RES, 2 DI near IEC header!  .. but in any case, I will follow your notes,  build one, and see how it goes with UNO or MEGA .. btw, I have built the standalone ArduinoFDC in the past just for fun,  and it worked great .. Thanks again David

**Comment by dhansel** [comment]
The TRN/RES/DI near the header are not really necessary. I added them to isolate the CLK/DATA/DATA pins on the Arduino
from the bus when the Arduino is not powered so it doesn't power itself backwards through those pins. It's not generally
a problem but if you turn on the computer before you power the Arduino then the Arduino may get into some weird state
because of the back-powering.

Note that you can actually also do without the 74HC125. It's not documented, but if you change the PIN_CTRL setting
in IECFDCMega.ino to "0xFF" (i.e. not set) then the library will do its best to work around the ATN/DATA timing issues.
It's not perfect and you may get a DEVICE NOT PRESENT error in some cases where you otherwise wouldn't
(e.g. if the device is busy formatting and you try to read the status channel). But it simplifies things for a first test.
All you need is to connect the Arduino to the IEC bus and the disk drive, no other components required.

Once you get that working you can add the 74HC125 to make things more stable (make sure to restore the PIN_CTRL 
setting after you add the chip).

**Comment by dhansel** [comment]
Forgot to say: Make sure to use the latest version of the IECDevice library, as I had to fix a few things to get this 
configuration back into working order.

**Comment by gilmod714** [comment]
Thank you so much David for your help & guidance - Greatly appreciated! .. I will build it and post my progress here with pics, I'm just waiting on some additional parts to arrive soon .. 

just a quick note about compiling for UNO .. in the examples folder "IECFDC" .. the sketch does not compile:  "over 104% not enough memory" , even after disabling Epyx & Dolphin  .. the MEGA compiles & uploads ok though ..



Best regards,
Gilmo

**Comment by gilmod714** [comment]
Hello David,

I'm running into some troubles with a few things .. ICP1 & OCP1 are on different pins for the MEGA shield, also, I am trying to understand the function of  PIN_Ctrl A4 pin  ?! .. 

just a quick overview about my setup & what I'm trying to do :

I built a XUM1541-promicro-v8 DIY and it was tested working .. (slow but ok) with my original 1541 drive -- until the head went out..

I also built  XA1541 parallel cable connecting to an older PC with SPP-LPT1 -- also worked but slow .. I did not know about jiffydos at the time .. I would have re-flashed a new ROM on 1541 , but it's too late now ..  (it's next to impossible to find head replacement for those old 1541's)

I have tested with OpenCBM & VICE with real drive support .. so the the xum1541 interface is verified working ok and sees real IEC devices..

Now I'm adding to this set up the IECproject/FDC on MEGA2560 as you can see from the pics .. and trying to put it all together .. it's challenging I know, that's why it's fun ;)


![1](https://github.com/user-attachments/assets/765f2182-a73f-49b1-be40-998f17b745e3)
![2](https://github.com/user-attachments/assets/3a29a47f-b7d2-4219-904a-cb8d97641ff1)
![3](https://github.com/user-attachments/assets/c3c8533e-f36f-4779-aedb-d9424634ba39)
![4](https://github.com/user-attachments/assets/245c5ce8-3b82-43cb-9048-2efbc3ea526c)
![5](https://github.com/user-attachments/assets/40672357-fc0c-4ccf-8ea5-1b4af23957db)
![6](https://github.com/user-attachments/assets/9d871059-1a99-4b0f-9cc6-28f2bd061a25)
![setup (Medium)](https://github.com/user-attachments/assets/62089d99-97ee-4a5e-901d-e370f8d3bb59)

- don't mind my 1541 PCB , I modified the power section for filtered DC power ATX PSU ..

**Comment by dhansel** [comment]
I don't quite understand your screen shot of the ArduinoFDC.cpp file. It looks like you moved the pin
assignments for Arduino MEGA into the section for the Arduino UNO. If you look down in the file
to the Arduino MEGA section (#elif defined(__AVR_ATmega2560__)) then you should find the pin
assignments matching the shield. You should not have to make any changes to the ArduinoFDC.cpp
file.

I have not tested the IECFDC sketch with the XUM1541. Interesting test case, I may get around to
that later today. I suspect that the XUM interface uses commands like M-R (memory read), 
M-W (memory write) and M-E (memory execute) to access the drive on a low level. That is not supported 
with this library. To support that you would need to full-on emulate the drive processor, memory and
everything. That is well beyond the scope of this project.

I'm a bit confused about what you are trying to do with this. The IECFDC example is meant to allow
a C64 to access a PC floppy drive. It looks to me like you are trying to connect a PC floppy drive to
a PC via two unnecessary conversions (PC ---XUM---> IEC bus --- IECFDC---> PC floppy). Why not
install the floppy drive directly in the PC or use a USB floppy drive?

**Comment by gilmod714** [comment]
Ahh yes, you are right about the pin-assignment for MEGA, I didn't see the section for the MEGA pinouts - my apologies

=== Why not install the floppy drive directly in the PC or use a USB floppy drive? ===
VICE Emulator will not see a real 3.5" PC drive ..  I'm trying to make VICE connect to a real IEC hardware (ex. 1541 .. or in this case a replacement/emulated 1581 3.5" on the IEC bus) .. when I used the XUM1541 , I was able to connect my 1541 and run games from my c64 disks when my 1541 worked .. also, building this project , I will be able to connect to my actual C64 but with 3.5" drive - when I get my c64 back (It's not with me right now),  as what this library is intended to do like u said .. later on, I can possibly convert my 5.25 c64 disks to 3.5 format and archive them properly .. I can also use my c64 disks with an emulator on the PC too .. so, i'm thinking multi-purpose/multi-goals here ;)

Thanks again David ..

![vice](https://github.com/user-attachments/assets/67c6eb65-989f-45b7-adcc-cdf07bf3d625)

**Comment by dhansel** [comment]
Oh, forgot to mention, the function of the CTRL pin is described in the "timing" section here:
https://github.com/dhansel/IECDevice?tab=readme-ov-file#timing-considerations
It controls the 74HC125 whose purpose is to pull DATA low as soon as ATN goes low. That way the 
bus master can see the drive is attached even if the Arduino is currently busy writing to disk and can't 
respond to ATN. Once the Arduino is not busy anymore it will disable the 74HC125 via the CTRL pin
and take control of the DATA signal (very similar to what a 1541 does).

I just did some tests here and found that I can successfully get the device status via my ZoomFloppy
(XUM relative) adapter. So the communication works. However, as expected, copying files via the XUM
does not work since the CBMtools programs attempt to upload and execute special code to drive RAM,
which will only work with a real 1541.

**Comment by gilmod714** [comment]
oh ok I see ... I know the 1541 has it's own 6502 processor and sram etc .. so I guess copying disks will not work, but reading/formatting disks should be ok , right?

**Comment by dhansel** [comment]
It really depends on what the program doing the reading or initiating the format does. If it expects a full 1541
and uses M-W/M-E commands then it won't work, otherwise it may.

**Comment by gilmod714** [comment]
ok, thanks David .. it is still fun to tinker and explore the possibilities :)

UPDATE:
I have changed the plan and adapted this disk ][ drive for 1541 controller (it works!) .. I posted my project on lemon64, thx for your support and guidance David :)

https://www.lemon64.com/forum/viewtopic.php?t=86143
